### PR TITLE
kqueue: make changelist and eventlist of kevent share the array

### PR DESCRIPTION
**Pull request by panjf2000**
The array used by the changelist of the kevent() can be reused by the eventlist, We should make use of this feature to eliminate the extra overhead of array allocation.

> The same array may be used for the changelist and eventlist.

It's cited from the man pages of freebsd/macos/dragonflybsd/netbsd/openbsd

### References

- [FreeBSD kqueue](https://man.freebsd.org/cgi/man.cgi?kqueue)
- [macOS kqueue](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/kqueue.2.html)
- [DragonFlyBSD kqueue](https://leaf.dragonflybsd.org/cgi/web-man?command=kqueue&section=2)
- [NetBSD](https://man.netbsd.org/kqueue.2)
- [OpenBSD](https://man.openbsd.org/kqueue.2)

### CODE REVIEW

**Comment by panjf2000** [comment]
Ping @azat

**Comment by panjf2000** [comment]
Got a minute for this? @azat

**Comment by panjf2000** [comment]
While you're here, please also take a look at this PR.

**Comment by panjf2000** [comment]
While you're here, please also take a look at this PR. @azat

**Comment by panjf2000** [comment]
> While you're here, please also take a look at this PR. @azat

Ping @azat

**Comment by panjf2000** [comment]
> > While you're here, please also take a look at this PR. @azat
> 
> Ping @azat

Ping @azat

File: kqueue.c [hunk 1]
```diff
@@ -420,9 +384,9 @@ kq_dispatch(struct event_base *base, struct timeval *tv)
 	}
 
 	if (res == kqop->events_size) {
-		/* We used all the events space that we have. Maybe we should
-		   make it bigger. */
-		kq_grow_events(kqop, kqop->events_size * 2);
+		/* We've used up all the events space in this round.
+		 * Try to double the size of the eventlist for next round. */
+		kq_grow_events(kqop);
```
Comment (azat) [inline kqueue.c:389]: I guess we can check exit code here as well

File: kqueue.c [hunk 2]
```diff
@@ -211,65 +228,32 @@ kq_build_changes_list(const struct event_changelist *changelist,
...
-			}
-			kqop->changes = newchanges;
-			kqop->changes_size = newsize;
-		}
 		if (in_ch->read_change) {
-			out_ch = &kqop->changes[n_changes++];
+			out_ch = &kqop->events[n_changes++];
 			kq_setup_kevent(out_ch, in_ch->fd, EVFILT_READ,
 			    in_ch->read_change);
+			if (n_changes == kqop->events_size && kq_grow_events(kqop)) {
```
Comment (azat) [inline kqueue.c:235]: Should we check for growing before access, i.e. before `out_ch` assignment?

File: kqueue.c [hunk 3]
```diff
@@ -278,47 +262,27 @@ kq_dispatch(struct event_base *base, struct timeval *tv)
...
+	 * while there is already one on the fly. */
+	events = kqop->events;
+	kqop->events = NULL;
 
 	EVBASE_RELEASE_LOCK(base, th_base_lock);
-
-	res = kevent(kqop->kq, changes, n_changes,
-	    events, kqop->events_size, ts_p);
-
+	res = kevent(kqop->kq, events, n_changes, events, kqop->events_size, ts_p);
```
Comment (azat) [inline kqueue.c:280]: Does passing `changelist==eventlist` safe for `kevent`? Please describe why

**Review by azat — COMMENTED** [review]
Sorry for such a huge delay! The major concern - is it safe?

File: kqueue.c (hunk 3 above)
Comment (panjf2000) [reply kqueue.c:280]: > The same array may be used for the changelist and eventlist.

cited from the man pages of freebsd/macos/dragonflybsd/netbsd/openbsd

File: kqueue.c (hunk 2 above)
Comment (panjf2000) [reply kqueue.c:235]: That would be needless. n_changes starts with zero and its value increment by one after one kevent in events is used (for every read or write event). That increased value will be checked each time we consume a kevent, which ensures that the size of events is doubled right after the old size is used up and makes sure there are always rooms for the next event and subsequent events.

File: kqueue.c (hunk 1 above)
Comment (panjf2000) [reply kqueue.c]: Done.

**Comment by panjf2000** [comment]
> Sorry for such a huge delay! The major concern - is it safe?

Cited from the man pages of freebsd/macos/dragonflybsd/netbsd/openbsd

> The same array may be used for the changelist and eventlist.

@azat

**Comment by panjf2000** [comment]
> > Sorry for such a huge delay! The major concern - is it safe?
> 
> Cited from the man pages of freebsd/macos/dragonflybsd/netbsd/openbsd
> 
> > The same array may be used for the changelist and eventlist.
> 
> This is also how `libuv` is using it.
> 
> @azat

@azat

**Comment by panjf2000** [comment]
> > Sorry for such a huge delay! The major concern - is it safe?
> 
> Cited from the man pages of freebsd/macos/dragonflybsd/netbsd/openbsd
> 
> > All changes contained in the changelist are applied before any pending events are read from the queue.
> > The same array may be used for the changelist and eventlist.
> 
> This is also how `libuv` is using it.
> 
> @azat

@azat

**Comment by panjf2000** [comment]
Also, I'm still expecting your reply to my email from a month ago.

**Review by azat — APPROVED** [review]

**Pull request merged**